internal/fs/local_file_test.go

```diff
@@ -177,29 +255,177 @@ func (t *LocalFileTest) RandomWritesToLocalFile() {
+	_, f2 := t.createLocalFile("implicitFoo/" + FileName2)
+
+	// Attempt to list implicit directory.
+	entries := t.readDirectory(path.Join(mntDir, "implicitFoo/"))
+
+	// Verify entries received successfully.
+	AssertEq(3, len(entries))
+	t.verifyLocalFileEntry(entries[0], "bar", 0)
+	t.verifyLocalFileEntry(entries[1], FileName, 0)
+	t.verifyLocalFileEntry(entries[2], FileName2, 0)
+	// Close the local files.
+	t.closeFileAndValidateObjectContents(f1, "implicitFoo/"+FileName, "")
+	t.closeFileAndValidateObjectContents(f2, "implicitFoo/"+FileName2, "")
+}
 
-	return
+func (t *LocalFileTest) TestRecursiveListingWithLocalFiles() {
+	/* Structure
+	mntDir/
+		- baseLocalFile 			--- file
+		- explicitFoo/ 				--- directory
+			- bar								--- file
+			- implicitLocalFile --- file
+		- explicitFoo/				--- directory
+			- explicitLocalFile --- file
+	*/
+
+	// Create implicit dir with 1 local file1 and 1 synced file.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				// File
+				"implicitFoo/bar": "",
+			}))
+	_, f1 := t.createLocalFile("implicitFoo/" + implicitLocalFileName)
+	// Create explicit dir with 1 local file.
+	AssertEq(
+		nil,
+		t.createObjects(
+			map[string]string{
+				"explicitFoo/": "",
+			}))
+	_, f2 := t.createLocalFile("explicitFoo/" + explicitLocalFileName)
+	// Create local file in mnt/ dir.
+	_, f3 := t.createLocalFile(FileName)
+
+	// Recursively list mntDir/ directory.
+	err := filepath.WalkDir(mntDir, func(path string, dir fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		// The object type is not directory.
+		if !dir.IsDir() {
+			return nil
+		}
+
+		objs, err := os.ReadDir(path)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		// Check if mntDir has correct objects.
+		if path == mntDir {
+			// numberOfObjects = 3
+			if len(objs) != 3 {
+				return fmt.Errorf("incorrect number of objects in the mntDir/")
+			}
+
+			if objs[0].Name() != "explicitFoo" || objs[0].IsDir() != true ||
+				objs[1].Name() != FileName || objs[1].IsDir() != false ||
+				objs[2].Name() != "implicitFoo" || objs[2].IsDir() != true {
+				return fmt.Errorf("listed incorrect object in mntDir/")
+			}
+		}
+
+		// Check if mntDir/explicitFoo/ has correct objects.
+		if path == mntDir+"/explicitFoo" {
+			// numberOfObjects = 1
+			if len(objs) != 1 {
+				return fmt.Errorf("incorrect number of objects in mntDir/explicitFoo/")
+			}
+
+			if objs[0].Name() != explicitLocalFileName || objs[0].IsDir() != false {
+				return fmt.Errorf("listed incorrect object in mntDir/explicitFoo/")
+			}
+		}
+
+		// Check if mntDir/implicitFoo/ has correct objects.
+		if path == mntDir+"/implicitFoo" {
+			// numberOfObjects = 2
+			if len(objs) != 2 {
+				return fmt.Errorf("incorrect number of objects in mntDir/implicitFoo/")
+			}
+
+			if objs[0].Name() != "bar" || objs[0].IsDir() != false ||
+				objs[1].Name() != implicitLocalFileName || objs[1].IsDir() != false {
```

Comment:
can we use Assert statements here similar to your other statements?

Reply:
done